### DISCUSSION TITLE
[RFC FS-1098] "Inline if lambda" attribute on function parameters of inline functions

### DISCUSSION THREAD

**Discussion by dsyme** [fsharp/fslang-design#549] [Language and core library RFC discussions]
Discussion thread for [RFC FS-1098 Inline if lambda attribute](https://github.com/fsharp/fslang-design/blob/master/RFCs/FS-1098-inline-if-lambda.md)

**Comment by Tarmil** [comment]
Will this also allow capturing byref types in the lambda? For example allowing something like this if `Array.iteri` had this attribute:

```fsharp
let copy (source: int[]) (target: Span<int>) =
    source |> Array.iteri (fun i x -> target.[i] <- x)
```

**Reply by baronfel** [reply]
Excellent question! I know there have been other language suggestions around making core library functions work with byreflike parameters, perhaps this would be an alternative pathway to enabling that?

**Reply by dsyme** [reply]
Yes, I have just added the ability to use this attribute on F#-defined delegate types, which can accept byrefs.

The optimizer know knows how to online and reduce invokes on these.

I will update the RFC tomorrow or Tuesday

**Reply by dsyme** [reply]
(F# functions can't accept byref parameters due to limitations of dotnet generics. But delegate type can if part of their signature )

**Reply by dsyme** [reply]
This works well for computation expressions, but may be awkward for bloke the above, since it may need new delegate type definition. It is what it is.

**Comment by krauthaufen** [comment]
Hey, I would have needed this quite often and the workarounds most often turned out to be code-duplication and less abstraction, so I would love to see that happen. 

I was wondering why we can't just inline all lambdas in functions being inline themselves when there is only one (syntactic) invocation? This rule would ensure that the code-size keeps small and cover a lot of cases.

**Reply by dsyme** [reply]
It's a good question.  I don't have a good answer :)
